Add PDF support

astashov opened the pull request:
First draft of the PDF support implementation.

For rendering PDF, we use PDF.js to render PDF in a WebView. To provide
PDF doc (which could be quite big) we stream it via the
Android<->JavaScript bridge, PDF.js requests chunks it needs, and we
serve them from the Android side through that bridge.

So, currently it works this way:

We check in WebFragment's `downloadListener` if the mime type is
`application/pdf`, and if yes - we download the PDF file, and then start
rendering the PDF loader (just an HTML file, that initializes PDF.js and
hooks up the bridge, and also exposes some additional API (e.g. for
zoom).

The bridge provides access to the downloaded file, via
`RandomAccessFile`. We need to support random access, therefore we can't
use `OkHttp`s `BodyResponse` and its streams, because it only allows to
read forward, and doesn't support going back.

The only way to make proper zoom is to rely on built-in `WebView`'s zoom.
So, I render PDF first with scale factor == 2, then set initial scale of
`WebView` to 1%, which makes it to max scale out so it fits the window size.

PDF.js render itself into a container you provide. Built-in PDF.js
`PdfViewer` uses the container's height to figure out what pages it
needs to load/render.  Since in Materialistic the `WebView`'s height
wraps the content, we should use window's height instead of container's
height for that. So, we inherit our custom `CustomPdfViewer` from
`PdfViewer` and override that functionality.

The only HTML/JS stuff I wrote is in `assets/pdf/index.html` and
`assets/pdf/script.js`. All other stuff in `assets/pdf/vendor/pdf.js/` is
taken from the PDF.js repo without any changes.

@hidroh, please take a look, and give me feedback on a general approach.
Please disregard the tests for now, I'll focus on them later, when the
functionality gonna be in place and correct.

Thanks!

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/DownloadModule.java:
```diff
@@ -0,0 +1,54 @@
+            .addHeader("Content-Type", mimeType)
+            .build();
+
+        mCallFactory.newCall(request).enqueue(new Callback() {
```
It'd be better to call `execute()` instead of `enqueue()` to avoid double callback, since the callback here only processes response on the same thread before relaying to the other callback.

astashov replied:
But the main thread calls `mDownloadModule.downloadFile`, if I replace `enqueue()` with `execute()` here, I'll get `NetworkOnMainThreadException`, right?

hidroh replied:
Yes you're right I didn't notice this method is to be called on main thread. I'm good with double callback for now, we can replace it with RxJava chain later.

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/DownloadModule.java:
```diff
@@ -0,0 +1,54 @@
+                byte[] data = new byte[1024];
+                int count;
+                while ((count = inputStream.read(data)) != -1) {
+                    outputStream.write(data, 0, count);
```
I prefer [Okio](https://github.com/square/okio), which is a more modern Java IO API, here for IO logic. The app already has Okio as a dependency. I'm not sure if you're familiar with it?

astashov replied:
I'm not, but will get familiar, no problem :)

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/DownloadModule.java:
```diff
@@ -0,0 +1,54 @@
+                }
+                inputStream.close();
+                outputStream.close();
+                callback.onSuccess(outputFile);
```
We can force this callback to be on main thread and hide the fact that we're operating on different threads from the outside world: `new Handler(Looper.getMainLooper()).post(() -> callback.onSuccess(outputFile))`. By convention, without any annotation I'd expect a method to be on main thread.

astashov replied:
Cool

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/DownloadModule.java:
```diff
@@ -0,0 +1,54 @@
+
+    public interface DownloadModuleCallback {
+        void onFailure(IOException e);
+        void onSuccess(File file) throws IOException;
```
I think it's better for thread safety to change callback interface to pass File's absolute path instead of File instance itself: `void onSuccess(String filePath)`. I also prefer not to throw `IOException` but to have this class eaten it up and fire an `onFailure()` callback instead.

astashov replied:
Makes sense

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/DownloadModule.java:
```diff
@@ -0,0 +1,54 @@
+        this.mCallFactory = callFactory;
+    }
+
+    void downloadFile(Context context, String url, String mimeType, DownloadModuleCallback callback) {
```
By Dagger convention, you'd need another level of abstraction here. A module is supposed to inject one or more dependencies, one of which will be `FileDownloader` (or a different name if you prefer) which has a method `downloadFile()`.

```
// Dagger will figure out where to get Context and Call.Factory to pass here thru other modules
@Provides
public FileDownloader provideFileDownloader(Context context, Call.Factory callFactory) {
  return new FileDownloader(context, callFactory);
}

public class FileDownloader {
  private final String mCacheDir;

  public FileDownloader(Context context, Call.Factory callFactory) {
    mCacheDir = context.getCacheDir().getPath(); // don't need to keep a reference to context after this
    // ...
  }

  public void downloadFile(String url, String mimeType, FileDownloaderCallback callback) {
  }
}

public interface FileDownloaderCallback { ... }
```

astashov replied:
Should I move this `FileDownloader` to `NetworkModule` then? Instead of creating `DownloadModule` just for the sake of this dependency?

astashov replied:
I.e. something like this? https://github.com/hidroh/materialistic/pull/1006/commits/56ca5d9433fa370613362536fc896ae93dbfaad7

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/WebFragment.java:
```diff
@@ -573,3 +623,69 @@ public void onError(String errorMessage) {
 }
+
+
+class PdfAndroidJavascriptBridge {
```
I don't normally put 2 classes at the root of same file, maybe make this a static inner class or move it to a separate file called `PdfAndroidJavascriptBridge.java`?

astashov replied:
I'll move to a static inner class

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/WebFragment.java:
```diff
@@ -288,7 +296,29 @@ protected void load() {
+            mPdfAndroidJavascriptBridge.cleanUp();
+            mWebView.removeJavascriptInterface("PdfAndroidJavascriptBridge");
+        }
+        if (isPdfRenderingSupported() && url.equals(PDF_LOADER_URL)) {
```
Let's replace `url.equals(PDF_LOADER_URL)` with `PDF_LOADER_URL.equalsIgnoreCase(url)` or `TextUtils.equals(url, PDF_LOADER_URL)` to avoid NPE if url is null?

astashov replied:
Okay, makes sense

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/WebFragment.java:
```diff
@@ -288,7 +296,29 @@ protected void load() {
+        reloadUrl(url, null);
+    }
+
+    private void reloadUrl(String url, File pdfFile) {
```
Can you annotate `pdfFile` argument with `@Nullable`: `@Nullable File pdfFile`? Also I think we need to skip some of logic below if `pdfFile` is null right?

astashov replied:
Yup

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/WebFragment.java:
```diff
@@ -404,17 +434,21 @@ public void onProgressChanged(android.webkit.WebView view, int newProgress) {
+            } else {
+                if (getActivity() == null) {
+                    return;
+                }
```
This early termination should still be top of the callback, without context (activity here) we can't really do nothing else.

astashov replied:
Right

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/WebFragment.java:
```diff
@@ -445,6 +479,9 @@ void setFullscreen(boolean isFullscreen) {
             reset();
+            if (mWebView.getUrl().equals(PDF_LOADER_URL)) {
+                mWebView.setInitialScale(1);
+            }
```
Let's remove the if condition here and force reset scale when we toggle between fullscreen & window mode?

astashov replied:
Sure, though apparently `setInitialScale` doesn't work, I should use `zoomOut` here.

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/WebFragment.java:
```diff
@@ -536,6 +573,19 @@ void onItemLoaded(@NonNull Item response) {
+
+            @Override
+            public void onSuccess(File file) throws IOException {
+                mWebView.post(() -> reloadUrl(PDF_LOADER_URL, file));
```
If we change the downloader to fire callback on main thread like in my other comment then we don't need to wrap this with `post()`.

astashov replied:
Yup

hidroh commented on app/src/main/assets/pdf/index.html:
```diff
@@ -0,0 +1,29 @@
+
+  <style>
+    body {
+      background-color: #808080;
```
I'm not sure how flexible PDF.js is with text color and background color? If it allows us to change colors we can inject them here to honor user's theme preference. But for a start we can go with black text on white background.

astashov replied:
Nah, this color is the background color of the document, and the pages go on top of it.
I.e.:

![image](https://user-images.githubusercontent.com/12795/33527762-e8c2b30a-d81b-11e7-96e2-803b834076f2.png)

It's somewhat similar to how Chrome renders PDFs:

![image](https://user-images.githubusercontent.com/12795/33527766-09441362-d81c-11e7-8a69-0d4f391780d8.png)

As for the actual document's text/background colors - it may be tricky, because a PDF doc may contain its own rules for the text/background colors. But since PDF.js just converts a PDF doc to a DOM tree with some CSS class names, it's probably possible to try to adjust them according to user's theme preference via custom CSS rules. Though since it's somewhat free-form, it again may be tricky to make it reliable and work consistently across different PDF documents..

hidroh commented on app/src/main/assets/pdf/vendor/pdf.js/pdf_viewer.js:
```diff
@@ -0,0 +1,5193 @@
+/* Copyright 2017 Mozilla Foundation
```
Is there a minified version of this file? We could put the minified version on release build and this version on debug build so that we can read the code during development but when we ship the size is as small as it can be.

Also I think it's better to add a version suffix to this file so we know what PDF.js version we're using, and add an attribution to Mozilla in `license.xml`.

hidroh commented on app/src/main/assets/pdf/vendor/pdf.js/pdf_viewer.css:
```diff
@@ -0,0 +1,338 @@
+/* Copyright 2014 Mozilla Foundation
```
Is there a minified version of this file? We could put the minified version on release build and this version on debug build so that we can read the code during development but when we ship the size is as small as it can be.

Also we can add version suffix to this file so we know what version of PDF.js we're using.

astashov commented:
@hidroh I addressed some of your feedback, still have a couple of questions though:

* https://github.com/hidroh/materialistic/pull/1006#discussion_r154839842
* https://github.com/hidroh/materialistic/pull/1006#discussion_r154839711

Please advice there :)

Thanks!

astashov commented:
Okay, I've added tests, hopefully fixed all your feedback (except https://github.com/hidroh/materialistic/pull/1006#discussion_r154519593), and made the build green. Please take another look.

hidroh commented:
Hey sorry for late reply, I didn't have time to look at your PR yet except to reply to your comment above, promise I'll look at it before weekend. Thanks for making the build green!

astashov commented:
> promise I'll look at it before weekend

Thanks!

hidroh commented on app/release/output.json:
```diff
@@ -0,0 +1 @@
+[{"outputType":{"type":"APK"},"apkInfo":{"type":"MAIN","splits":[],"versionCode":74},"path":"app-release.apk","properties":{"packageId":"io.github.hidroh.materialistic","split":"","minSdkVersion":"14"}}]
```
I think this file is not needed right?

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/data/FileDownloader.java:
```diff
@@ -0,0 +1,57 @@
+        mCallFactory.newCall(request).enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException e) {
+                callback.onFailure(e);
```
This needs to be wrapped with `new Handler(Looper.getMainLooper()).post(() -> callback.onFailure(e))` as well.

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/data/FileDownloader.java:
```diff
@@ -0,0 +1,57 @@
+                    sink.close();
+                    new Handler(Looper.getMainLooper()).post(() -> callback.onSuccess(outputFile.getPath()));
+                } catch (IOException e) {
+                    callback.onFailure(e);
```
Just forward to `onFailure(call, e)` here?

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/WebFragment.java:
```diff
@@ -445,6 +477,8 @@ void setFullscreen(boolean isFullscreen) {
         } else {
             reset();
+            // We'll zoom out until it returns false, which means it has min zoom level
+            while (mWebView.zoomOut()) {}
```
I hope this work as expected and we wouldn't end up in infinite loop here 😬

astashov replied:
According to API docs - it should, but you're right - it's damn dangerous if it doesn't! Maybe add some artificial limit - like do that 30 times, and then force break it.

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/WebFragment.java:
```diff
@@ -536,6 +570,19 @@ void onItemLoaded(@NonNull Item response) {
+        mFileDownloader.downloadFile(mItem.getUrl(), PDF_MIME_TYPE, new FileDownloader.FileDownloaderCallback() {
+            @Override
+            public void onFailure(IOException e) {
+            }
```
Should we now fall back to the flow where we prompt user to open it in another app? Or at least show a toast?

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/WebFragment.java:
```diff
@@ -572,4 +619,62 @@ public void onError(String errorMessage) {
+            }
+        }
+
+        public void finalize() throws Throwable {
```
Can you add an `@Override` here since it overrrides super class method.

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/WebFragment.java:
```diff
@@ -288,7 +292,31 @@ protected void load() {
+    @SuppressLint("AddJavascriptInterface")
+    private void reloadUrl(String url, @Nullable String pdfFilePath) {
+        if (mPdfAndroidJavascriptBridge != null) {
+            mPdfAndroidJavascriptBridge.cleanUp();
```
Shouldn't this be `finalize()` instead?

hidroh replied:
Also we should call it in `onDestroy()` as well, or else we only clean up the previously opened file when we need to load a new one.

astashov replied:
I thought it's considered a bad practice to call `finallize()` explicitly, and the GC gonna call it whenever it needs to. And here I want to clean it up before GC, because we don't really need the opened PDF file for the previous page anymore.

hidroh replied:
Ah ok got it, I forgot that `finalize()` is destructor in Java, so rarely use it. But if we call `cleanUp()` in `onDestroy()` it's more in our control to clean up when we close the article, you don't want to wait until it hits GC.

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/WebFragment.java:
```diff
@@ -572,4 +619,62 @@ public void onError(String errorMessage) {
+            mContext = context;
+            mFile = new File(filePath);
+            try {
+                mRandomAccessFile = new RandomAccessFile(mFile, "r");
```
Do you think it's better if we defer this instantiation to beginning of `getChunk()`?

astashov replied:
Probably won't make much difference, since it gonna happen several milliseconds later :) But okay, I'lll move it.

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/data/FileDownloader.java:
```diff
@@ -0,0 +1,57 @@
+        mCallFactory = callFactory;
+    }
+
+    public void downloadFile(String url, String mimeType, FileDownloaderCallback callback) {
```
Let's annotate this method with `@WorkerThread`.

hidroh commented on app/src/main/java/io/github/hidroh/materialistic/WebFragment.java:
```diff
@@ -572,4 +619,62 @@ public void onError(String errorMessage) {
+    static class PdfAndroidJavascriptBridge {
+        Context mContext;
+        File mFile;
+        @Nullable RandomAccessFile mRandomAccessFile;
```
Nit: The above fields can be private.

hidroh commented:
I did a little bit of testing but I can only get PDFs to render about 40-50% of the time, not sure what happened to the failed ones. The steps I took are:
- search for 'pdf', sort by popularity
- open the PDF file by tapping on each item in search results

Also I think we need to (these I can do after we merge):
- [ ] remove back/forward options in fullscreen since they don't really apply to PDF docs

astashov commented:
Could you paste links for some failed PDFs here, I'd take a look tonight.
Back/forward could actually apply, since we support links in PDFs, so if a PDF has a link to external site you can go back then to the PDF doc.

astashov commented:
Also, if we fail by some reason (can't read/parse PDF file, or can't download a file, etc) - I think it makes sense to fall back to the old dialog, which used to appear when you try to open a PDF file. What do you think?

hidroh commented:
Yes I suggested the same when I reviewed above.

Of the top 5 here, the first 3 won't render: https://hn.algolia.com/?query=pdf&sort=byPopularity&prefix=false&page=0&dateRange=all&type=story, you can get these when you search for 'pdf' in the app.

astashov commented:
First 2 rendered for me (I've tried in a Nexus 6 emulator), the 3rd one - don't.
I'll take a look tonight, thanks for pointing out!

astashov commented:
@hidroh please take a look again. I think I've addressed all your feedback, except the `finalize()` comment (https://github.com/hidroh/materialistic/pull/1006#discussion_r155860180) where I left a comment.

hidroh commented:
Looking good to me 👍 

There is a very minor UI issue with the progress bar, as it shows, hides, then shows again. I think the 1st show is for file download progress, and the 2nd show is for PDF rendering progress. The pause between the 2 sometimes can last a few seconds making me think that it fails to load. But we can address it after we merge this PR.

hidroh commented:
Closes #998

astashov commented:
> I think the 1st show is for file download progress, and the 2nd show is for PDF rendering progress. The pause between the 2 sometimes can last a few seconds making me think that it fails to load. But we can address it after we merge this PR.

I can take a look into fixing the progress bar this weekend, I agree it's far from ideal for the PDF docs.

Thanks for merging it!
